File: cursor_example/rmm_allocation_example.cpp
/*
 * RMM Allocation Strategy Example
 * 
 * This example demonstrates:
 * 1. How RMM handles small vs large allocations
 * 2. Different memory resource types
 * 3. When cudaFree is actually called
 */

#include <iostream>
#include <iomanip>
#include <sstream>
#include <vector>
#include <chrono>
#include <rmm/cuda_device.hpp>
#include <rmm/cuda_stream.hpp>
#include <rmm/device_uvector.hpp>
#include <rmm/device_buffer.hpp>
#include <rmm/mr/device/cuda_memory_resource.hpp>
#include <rmm/mr/device/pool_memory_resource.hpp>
#include <rmm/mr/device/binning_memory_resource.hpp>
#include <rmm/mr/device/fixed_size_memory_resource.hpp>
#include <rmm/mr/device/cuda_async_memory_resource.hpp>
#include <cuda_runtime_api.h>

// Helper to format bytes
std::string format_bytes(std::size_t bytes) {
    const char* units[] = {"B", "KB", "MB", "GB"};
    int unit = 0;
    double size = static_cast<double>(bytes);
    while (size >= 1024.0 && unit < 3) {
        size /= 1024.0;
        unit++;
    }
    std::ostringstream oss;
    oss << std::fixed << std::setprecision(2) << size << " " << units[unit];
    return oss.str();
}

// Example 1: Direct CUDA allocations (immediate cudaFree)
void example_direct_cuda_allocations() {
    std::cout << "\n" << std::string(60, '=') << "\n";
    std::cout << "Example 1: Direct CUDA Allocations (cuda_memory_resource)\n";
    std::cout << "Note: Each deallocate() immediately calls cudaFree()\n";
    std::cout << std::string(60, '=') << "\n";
    
    rmm::mr::cuda_memory_resource cuda_mr;
    rmm::cuda_stream stream;
    
    // Small allocations
    std::vector<rmm::device_uvector<char>> small_allocs;
    std::cout << "\nAllocating small buffers (1KB, 4KB, 16KB):\n";
    for (auto size : {1024, 4096, 16384}) {
        small_allocs.emplace_back(size, stream, cuda_mr);
        std::cout << "  Allocated " << format_bytes(size) 
                  << " at pointer: " << static_cast<void*>(small_allocs.back().data()) << "\n";
    }
    
    // Large allocations
    std::vector<rmm::device_uvector<char>> large_allocs;
    std::cout << "\nAllocating large buffers (1MB, 10MB, 100MB):\n";
    for (auto size : {1024*1024, 10*1024*1024, 100*1024*1024}) {
        large_allocs.emplace_back(size, stream, cuda_mr);
        std::cout << "  Allocated " << format_bytes(size) 
                  << " at pointer: " << static_cast<void*>(large_allocs.back().data()) << "\n";
    }
    
    stream.synchronize();
    std::cout << "\nWhen these buffers go out of scope, cudaFree() is called immediately.\n";
}

// Example 2: Pool allocations (delayed cudaFree)
void example_pool_allocations() {
    std::cout << "\n" << std::string(60, '=') << "\n";
    std::cout << "Example 2: Pool Memory Resource\n";
    std::cout << "Note: Memory goes to pool free list, NO immediate cudaFree()\n";
    std::cout << std::string(60, '=') << "\n";
    
    rmm::mr::cuda_memory_resource upstream_mr;
    
    // Create a pool with initial size of 256MB, max 1GB
    constexpr std::size_t initial_pool = 256 * 1024 * 1024;
    constexpr std::size_t max_pool = 1024 * 1024 * 1024;
    
    rmm::mr::pool_memory_resource<rmm::mr::cuda_memory_resource> 
        pool_mr(&upstream_mr, initial_pool, max_pool);
    
    std::cout << "\nPool created:\n";
    std::cout << "  Initial pool size: " << format_bytes(initial_pool) << "\n";
    std::cout << "  Maximum pool size: " << format_bytes(max_pool) << "\n";
    std::cout << "  Current pool size: " << format_bytes(pool_mr.pool_size()) << "\n";
    
    rmm::cuda_stream stream;
    
    // Allocate many small buffers
    std::cout << "\nAllocating 10 buffers of varying sizes from pool:\n";
    std::vector<rmm::device_uvector<char>> buffers;
    for (int i = 0; i < 10; ++i) {
        auto size = (i + 1) * 1024 * 1024;  // 1MB to 10MB
        buffers.emplace_back(size, stream, pool_mr);
        std::cout << "  Buffer " << i+1 << ": " << format_bytes(size) 
                  << " (pool size: " << format_bytes(pool_mr.pool_size()) << ")\n";
    }
    
    stream.synchronize();
    std::cout << "\nCurrent pool size after allocations: " 
              << format_bytes(pool_mr.pool_size()) << "\n";
    
    std::cout << "\nFreeing all buffers...\n";
    buffers.clear();  // Returns memory to pool free list
    
    std::cout << "Pool size after deallocation: " 
              << format_bytes(pool_mr.pool_size()) << "\n";
    std::cout << "Note: cudaFree() NOT called yet - memory remains in pool!\n";
    std::cout << "cudaFree() will only be called when pool_mr is destroyed.\n";
}

// Example 3: Binning memory resource (optimized for different sizes)
void example_binning_allocations() {
    std::cout << "\n" << std::string(60, '=') << "\n";
    std::cout << "Example 3: Binning Memory Resource\n";
    std::cout << "Routes allocations to different resources based on size\n";
    std::cout << std::string(60, '=') << "\n";
    
    rmm::mr::cuda_memory_resource upstream_mr;
    
    // Create binning resource with fixed-size bins for small allocations
    // Bins: 1KB, 4KB, 16KB, 64KB, 256KB, 1MB, 4MB
    rmm::mr::binning_memory_resource<rmm::mr::cuda_memory_resource> 
        binning_mr(&upstream_mr, 10, 22);  // 2^10 = 1KB to 2^22 = 4MB
    
    // Add a pool for allocations larger than 4MB
    auto pool_for_large = std::make_unique<
        rmm::mr::pool_memory_resource<rmm::mr::cuda_memory_resource>>(
            &upstream_mr, 100 * 1024 * 1024, 500 * 1024 * 1024);
    binning_mr.add_bin(4 * 1024 * 1024 + 1, pool_for_large.get());
    
    rmm::cuda_stream stream;
    
    std::cout << "\nAllocation size routing:\n";
    std::cout << "  Small (< 1KB):      Fixed-size bin\n";
    std::cout << "  Medium (1KB-4MB):   Fixed-size bins\n";
    std::cout << "  Large (> 4MB):      Pool memory resource\n";
    
    // Small allocations - use fixed-size bins
    std::cout << "\nSmall allocations (use fixed-size bins):\n";
    std::vector<rmm::device_uvector<char>> small_buffers;
    for (auto size : {512, 2048, 8192, 32768, 131072}) {
        small_buffers.emplace_back(size, stream, binning_mr);
        std::cout << "  " << format_bytes(size) << " -> Fixed-size bin\n";
    }
    
    // Large allocations - use pool
    std::cout << "\nLarge allocations (use pool resource):\n";
    std::vector<rmm::device_uvector<char>> large_buffers;
    for (auto size : {5*1024*1024, 20*1024*1024, 50*1024*1024}) {
        large_buffers.emplace_back(size, stream, binning_mr);
        std::cout << "  " << format_bytes(size) << " -> Pool resource\n";
    }
    
    stream.synchronize();
    
    std::cout << "\nWhen buffers are freed:\n";
    std::cout << "  Small buffers: Returned to fixed-size bin free lists\n";
    std::cout << "  Large buffers: Returned to pool free list\n";
    std::cout << "  cudaFree() only called when resources are destroyed\n";
}

// Example 4: Fixed-size memory resource (very fast for same-sized allocations)
void example_fixed_size_allocations() {
    std::cout << "\n" << std::string(60, '=') << "\n";
    std::cout << "Example 4: Fixed-Size Memory Resource\n";
    std::cout << "O(1) allocation for fixed-size blocks\n";
    std::cout << std::string(60, '=') << "\n";
    
    rmm::mr::cuda_memory_resource upstream_mr;
    
    // Create fixed-size resource for 64KB blocks
    constexpr std::size_t block_size = 64 * 1024;
    rmm::mr::fixed_size_memory_resource<rmm::mr::cuda_memory_resource>
        fixed_mr(&upstream_mr, block_size, 128);  // 128 blocks pre-allocated
    
    std::cout << "\nFixed-size resource:\n";
    std::cout << "  Block size: " << format_bytes(block_size) << "\n";
    std::cout << "  Blocks pre-allocated: 128\n";
    std::cout << "  Total pre-allocated: " << format_bytes(block_size * 128) << "\n";
    
    rmm::cuda_stream stream;
    
    // Allocate many blocks of the same size
    std::cout << "\nAllocating 20 blocks (should be very fast - O(1)):\n";
    auto start = std::chrono::high_resolution_clock::now();
    
    std::vector<rmm::device_uvector<char>> blocks;
    for (int i = 0; i < 20; ++i) {
        blocks.emplace_back(block_size, stream, fixed_mr);
    }
    
    stream.synchronize();
    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
    
    std::cout << "  Allocated 20 blocks in " << duration.count() << " microseconds\n";
    std::cout << "  Average: " << (duration.count() / 20.0) << " microseconds per allocation\n";
    
    std::cout << "\nFreeing all blocks (also O(1)):\n";
    start = std::chrono::high_resolution_clock::now();
    blocks.clear();
    end = std::chrono::high_resolution_clock::now();
    duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
    
    std::cout << "  Freed 20 blocks in " << duration.count() << " microseconds\n";
    std::cout << "  Memory returned to fixed-size free list\n";
    std::cout << "  cudaFree() only called when fixed_mr is destroyed\n";
}

// Example 5: CUDA async memory resource (uses CUDA's built-in pool)
void example_cuda_async_allocations() {
    std::cout << "\n" << std::string(60, '=') << "\n";
    std::cout << "Example 5: CUDA Async Memory Resource\n";
    std::cout << "Uses CUDA driver's built-in memory pool\n";
    std::cout << std::string(60, '=') << "\n";
    
    // Create async resource using 50% of free GPU memory
    rmm::mr::cuda_async_memory_resource async_mr{
        rmm::percent_of_free_device_memory(50)};
    
    rmm::cuda_stream stream;
    
    std::cout << "\nCUDA async memory resource uses CUDA driver's memory pool.\n";
    std::cout << "Allocations are stream-ordered and very efficient.\n\n";
    
    // Allocate buffers of different sizes
    std::vector<rmm::device_uvector<float>> buffers;
    std::cout << "Allocating buffers:\n";
    for (auto size_mb : {1, 5, 10, 25}) {
        auto size = size_mb * 1024 * 1024 / sizeof(float);
        buffers.emplace_back(size, stream, async_mr);
        std::cout << "  " << size_mb << " MB buffer (" << size << " floats)\n";
    }
    
    stream.synchronize();
    
    std::cout << "\nWhen freed, memory returns to CUDA driver's pool.\n";
    std::cout << "No cudaFree() called - CUDA driver manages the pool.\n";
    buffers.clear();
}

int main(int argc, char** argv) {
    std::cout << "\n";
    std::cout << "╔══════════════════════════════════════════════════════════╗\n";
    std::cout << "║     RMM Allocation Strategy Examples                     ║\n";
    std::cout << "║     Understanding Large vs Small Allocations             ║\n";
    std::cout << "║     When cudaFree() is Actually Called                   ║\n";
    std::cout << "╚══════════════════════════════════════════════════════════╝\n";
    
    try {
        // Show GPU info
        auto device_id = rmm::get_current_cuda_device();
        auto [free_mem, total_mem] = rmm::available_device_memory();
        
        // Get device name
        cudaDeviceProp prop;
        cudaGetDeviceProperties(&prop, device_id.value());
        
        std::cout << "\nGPU Information:\n";
        std::cout << "  Device ID: " << device_id.value() << "\n";
        std::cout << "  Device Name: " << prop.name << "\n";
        std::cout << "  Total Memory: " << format_bytes(total_mem) << "\n";
        std::cout << "  Free Memory: " << format_bytes(free_mem) << "\n";
        
        // Run examples
        example_direct_cuda_allocations();
        example_pool_allocations();
        example_binning_allocations();
        example_fixed_size_allocations();
        example_cuda_async_allocations();
        
        std::cout << "\n" << std::string(60, '=') << "\n";
        std::cout << "Summary:\n";
        std::cout << "  1. cuda_memory_resource: Immediate cudaFree()\n";
        std::cout << "  2. pool_memory_resource: Delayed cudaFree() (on destruction)\n";
        std::cout << "  3. binning_memory_resource: Routes to appropriate resource\n";
        std::cout << "  4. fixed_size_memory_resource: O(1) for fixed sizes\n";
        std::cout << "  5. cuda_async_memory_resource: Uses CUDA driver pool\n";
        std::cout << std::string(60, '=') << "\n";
        
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }
    
    return 0;
}

File: cursor_example/rmm_allocation_example.py
#!/usr/bin/env python3
"""
RMM Allocation Strategy Example (Python)

This example demonstrates:
1. How RMM handles small vs large allocations
2. Different memory resource types
3. When cudaFree is actually called (through Python's RMM interface)
"""

import rmm
import rmm.mr
import numpy as np
import time

def format_bytes(bytes_val):
    """Format bytes to human-readable string"""
    for unit in ['B', 'KB', 'MB', 'GB']:
        if bytes_val < 1024.0:
            return f"{bytes_val:.2f} {unit}"
        bytes_val /= 1024.0
    return f"{bytes_val:.2f} TB"

def example_direct_cuda_allocations():
    """Example 1: Direct CUDA allocations (immediate cudaFree)"""
    print("\n" + "="*60)
    print("Example 1: Direct CUDA Allocations (CudaMemoryResource)")
    print("Note: Each deallocate() immediately calls cudaFree()")
    print("="*60)
    
    # Use default CUDA memory resource
    cuda_mr = rmm.mr.CudaMemoryResource()
    rmm.mr.set_current_device_resource(cuda_mr)
    
    print("\nAllocating small buffers (1KB, 4KB, 16KB):")
    small_buffers = []
    for size in [1024, 4096, 16384]:
        buf = rmm.DeviceBuffer(size=size)
        small_buffers.append(buf)
        print(f"  Allocated {format_bytes(size)} at pointer: {buf.ptr}")
    
    print("\nAllocating large buffers (1MB, 10MB, 100MB):")
    large_buffers = []
    for size in [1024*1024, 10*1024*1024, 100*1024*1024]:
        buf = rmm.DeviceBuffer(size=size)
        large_buffers.append(buf)
        print(f"  Allocated {format_bytes(size)} at pointer: {buf.ptr}")
    
    print("\nWhen these buffers are deleted, cudaFree() is called immediately.")
    del small_buffers, large_buffers

def example_pool_allocations():
    """Example 2: Pool allocations (delayed cudaFree)"""
    print("\n" + "="*60)
    print("Example 2: Pool Memory Resource")
    print("Note: Memory goes to pool free list, NO immediate cudaFree()")
    print("="*60)
    
    # Create upstream resource
    upstream_mr = rmm.mr.CudaMemoryResource()
    
    # Create pool with initial 256MB, max 1GB
    # Note: sizes can be strings like "256MB" or integers
    pool_mr = rmm.mr.PoolMemoryResource(
        upstream_mr,  # First argument is the upstream resource
        initial_pool_size="256MB",
        maximum_pool_size="1GB"
    )
    
    print(f"\nPool created:")
    print(f"  Initial pool size: 256MB")
    print(f"  Maximum pool size: 1GB")
    
    rmm.mr.set_current_device_resource(pool_mr)
    
    print("\nAllocating 10 buffers of varying sizes from pool:")
    buffers = []
    for i in range(10):
        size = (i + 1) * 1024 * 1024  # 1MB to 10MB
        buf = rmm.DeviceBuffer(size=size)
        buffers.append(buf)
        print(f"  Buffer {i+1}: {format_bytes(size)}")
    
    print("\nFreeing all buffers...")
    del buffers
    
    print("Note: cudaFree() NOT called yet - memory remains in pool!")
    print("cudaFree() will only be called when pool_mr is destroyed.")

def example_binning_allocations():
    """Example 3: Binning memory resource (optimized for different sizes)"""
    print("\n" + "="*60)
    print("Example 3: Binning Memory Resource")
    print("Routes allocations to different resources based on size")
    print("="*60)
    
    upstream_mr = rmm.mr.CudaMemoryResource()
    
    # Create binning resource
    # Small allocations will use fixed-size bins
    # Large allocations will use upstream resource
    binning_mr = rmm.mr.BinningMemoryResource(
        upstream_mr,  # First argument is the upstream resource
        min_size_exponent=10,  # 2^10 = 1KB
        max_size_exponent=22   # 2^22 = 4MB
    )
    
    # For allocations > 4MB, use the pool
    # Note: Python API may not expose add_bin directly, so we use a simplified version
    rmm.mr.set_current_device_resource(binning_mr)
    
    print("\nAllocation size routing:")
    print("  Small (< 1KB):      Fixed-size bin")
    print("  Medium (1KB-4MB):   Fixed-size bins")
    print("  Large (> 4MB):      Upstream resource (or configured pool)")
    
    print("\nSmall allocations (use fixed-size bins):")
    small_buffers = []
    for size in [512, 2048, 8192, 32768, 131072]:
        buf = rmm.DeviceBuffer(size=size)
        small_buffers.append(buf)
        print(f"  {format_bytes(size)} -> Fixed-size bin")
    
    print("\nLarge allocations:")
    large_buffers = []
    for size in [5*1024*1024, 20*1024*1024]:
        buf = rmm.DeviceBuffer(size=size)
        large_buffers.append(buf)
        print(f"  {format_bytes(size)} -> Pool/upstream resource")
    
    print("\nWhen buffers are freed:")
    print("  Small buffers: Returned to fixed-size bin free lists")
    print("  Large buffers: Returned to pool/upstream")
    print("  cudaFree() only called when resources are destroyed")
    
    del small_buffers, large_buffers

def example_reinitialize_pool():
    """Example 4: Using reinitialize for easy pool setup"""
    print("\n" + "="*60)
    print("Example 4: Easy Pool Setup with reinitialize()")
    print("="*60)
    
    # Simple way to set up RMM with a pool
    rmm.reinitialize(
        pool_allocator=True,
        initial_pool_size="256MB",
        maximum_pool_size="1GB"
    )
    
    print("\nRMM initialized with pool:")
    print("  Pool allocator: Enabled")
    print("  Initial pool size: 256MB")
    print("  Maximum pool size: 1GB")
    
    print("\nAllocating buffers (all use the pool):")
    buffers = []
    start_time = time.time()
    
    for i in range(20):
        size = (i % 5 + 1) * 1024 * 1024  # 1MB to 5MB
        buf = rmm.DeviceBuffer(size=size)
        buffers.append(buf)
    
    alloc_time = time.time() - start_time
    print(f"  Allocated {len(buffers)} buffers in {alloc_time*1000:.2f} ms")
    
    num_buffers = len(buffers)
    start_time = time.time()
    del buffers
    free_time = time.time() - start_time
    print(f"  Freed {num_buffers} buffers in {free_time*1000:.2f} ms")
    
    print("\nNote: Memory is in the pool, ready for reuse!")
    print("No cudaFree() called until RMM is finalized or pool is destroyed.")

def example_statistics_tracking():
    """Example 5: Track allocation statistics"""
    print("\n" + "="*60)
    print("Example 5: Statistics Resource Adaptor")
    print("Track memory allocation patterns")
    print("="*60)
    
    upstream_mr = rmm.mr.CudaMemoryResource()
    
    # Wrap with statistics tracking
    # StatisticsResourceAdaptor takes upstream as first argument
    stats_mr = rmm.mr.StatisticsResourceAdaptor(upstream_mr)
    rmm.mr.set_current_device_resource(stats_mr)
    
    print("\nAllocating buffers with statistics tracking:")
    buffers = []
    
    for i, size in enumerate([1024, 4096, 1024*1024, 10*1024*1024]):
        buf = rmm.DeviceBuffer(size=size)
        buffers.append(buf)
        print(f"  Buffer {i+1}: {format_bytes(size)}")
    
    # Get statistics (Statistics object with attributes, not a dict)
    stats = stats_mr.allocation_counts
    print("\nAllocation Statistics:")
    print(f"  Current bytes allocated: {format_bytes(stats.current_bytes)}")
    print(f"  Current allocation count: {stats.current_count}")
    print(f"  Peak bytes allocated: {format_bytes(stats.peak_bytes)}")
    print(f"  Peak allocation count: {stats.peak_count}")
    print(f"  Total bytes allocated: {format_bytes(stats.total_bytes)}")
    print(f"  Total allocation count: {stats.total_count}")
    
    print("\nFreeing buffers...")
    del buffers
    
    stats = stats_mr.allocation_counts
    print("\nStatistics after freeing:")
    print(f"  Current bytes allocated: {format_bytes(stats.current_bytes)}")
    print(f"  Current allocation count: {stats.current_count}")
    print(f"  Peak bytes was: {format_bytes(stats.peak_bytes)}")

def main():
    print("\n")
    print("╔══════════════════════════════════════════════════════════╗")
    print("║     RMM Allocation Strategy Examples (Python)            ║")
    print("║     Understanding Large vs Small Allocations             ║")
    print("║     When cudaFree() is Actually Called                   ║")
    print("╚══════════════════════════════════════════════════════════╝")
    
    try:
        # Show GPU info
        import rmm._cuda.gpu as gpu
        device_id = gpu.getDevice()
        print(f"\nGPU Information:")
        print(f"  Device ID: {device_id}")
        
        # Run examples
        example_direct_cuda_allocations()
        example_pool_allocations()
        example_reinitialize_pool()
        example_statistics_tracking()
        
        print("\n" + "="*60)
        print("Summary:")
        print("  1. CudaMemoryResource: Immediate cudaFree()")
        print("  2. PoolMemoryResource: Delayed cudaFree() (on destruction)")
        print("  3. BinningMemoryResource: Routes to appropriate resource")
        print("  4. reinitialize(): Easy way to set up pool allocation")
        print("  5. StatisticsResourceAdaptor: Track allocation patterns")
        print("="*60)
        print("\n✅ All examples completed successfully!")
        
    except Exception as e:
        print(f"\n❌ Error: {e}")
        import traceback
        traceback.print_exc()
        return 1
    
    return 0

if __name__ == "__main__":
    exit(main())

